File: 2304-minimum-path-cost-in-a-grid/2304-minimum-path-cost-in-a-grid.py
class Solution:
    def minPathCost(self, gd: List[List[int]], mc: List[List[int]]) -> int:
        n = len(gd)
        m = len(gd[0])
        dp = [[-1 for i in range(m)]for j in range (n)]
        def go(r,c):
            if dp[r][c]!=-1:
                return dp[r][c]
            if(r==0):
                dp[r][c] = gd[r][c]
                return dp[r][c]
            x = 1e18
            for i in range (m):
                x = min(x, mc[gd[r-1][i]][c]+go(r-1,i))
            x+=gd[r][c]
            dp[r][c]=x
            return dp[r][c]
        ans = 1e18
        for i in range (m):
            ans = min(ans,go(n-1,i))
        return ans
        

File: 2304-minimum-path-cost-in-a-grid/2304-minimum-path-cost-in-a-grid.cpp
class Solution {
public:
    vector<vector<int>> gd,mc;
    int dp[51][51];
    int go(int row,int col){
        if(dp[row][col]!=-1) return dp[row][col];
        if(row==0){
            return dp[row][col] = gd[row][col];
        }
        int x = INT_MAX;
        for(int i=0;i<gd[0].size();++i){
            x = min(x,mc[gd[row-1][i]][col]+go(row-1,i));
        }
        x+=gd[row][col];
        return dp[row][col] = x;
    }
    int minPathCost(vector<vector<int>>& g, vector<vector<int>>& m) {
        gd = g;
        mc = m;
        int ans = INT_MAX;
        for(int i=0;i<51;++i)
            for(int j=0;j<51;++j)
                dp[i][j]=-1;
        for(int i=0;i<gd[0].size();++i){
            ans = min(ans,go(gd.size()-1,i));
        }       
        return ans ;
    }
};